src/luascript.cpp

```diff
@@ -9738,16 +9739,17 @@ int LuaScriptInterface::luaPlayerGetStorageValue(lua_State* L)
 	int32_t value;
 	if (player->getStorageValue(key, value)) {
 		lua_pushnumber(L, value);
+	} else if (lua_isnil(L, 3)) {
+		lua_pushnil(L);
```

Comment:
Idea: take a default value as third parameter, but when it is omitted, log a warning that the default will change from -1 to nil in the future and the user must use the third parameter to keep the previous behavior or silence the warning.

Reply:
There are too many scripts with logic based on 'default storage = -1'.
Printing error won't help, people can't rewrite all their scripts just because it would be better to return `nil`.

I would make it configurable in `config.lua`. Something like `defaultStorageValueNil` with `false` as default value [returns -1].
If someone decides to make new clean datapack, he can change it in `config.lua` and write all scripts with new `nil` logic.

Reply:
It's a complicated issue, I don't think it's the right time for this change yet. I've been testing with a default value of nil, and it's inconvenient. We have to add a lot of extra checks everywhere.

```lua
if player:getStorageValue(1) > os.time() then
-- Error
end
```

Solution:
```lua
local value = player:getStorageValue(1)
if value and value  > os.time() then
-- Ok
end
-- OR
if player:getStorageValue(1, 0)  > os.time() then
-- Ok
end
```

We all believe that this method will return a number.

I think what we can use is:
```lua
if player:hasStorageKey(1) then
-- get value
end
```
However, here we would be performing two searches in the storage map.

I really wouldn't like to have to write a default value everywhere, and otherwise check if the value is nil before performing arithmetic operations.

Reply:
@MillhioreBT 

As I said. Default configuration should make `getStorage` return `-1` as it's now.
We just add default value for new scripts.

> It's a complicated issue, I don't think it's the right time for this change yet. I've been testing with a default value of nil, and it's inconvenient. We have to add a lot of extra checks everywhere.
> 
> ```lua
> if player:getStorageValue(1) > os.time() then
> -- Error
> end
> ```

When you start using `nil logic`, you also start using own `default value` in 99% of code.
In most cases it's 0 ex. start counting killed creatures from 0 in task system.
In other common case, you check, if player has some storage (did quest) - in this case you use `hasStorageValue` function, not `getStorageValue`. Any value set to given storage means that player did quest. `hasStorageValue` makes it obvious for anyone that it's true/false value. Much easier to understand than `getStorageValue(X) > 0`.

Your script in new system would be:
```
if player:getStorageValue(1, 0) > os.time() then
   -- no error, just your logic
end
```

>
> I really wouldn't like to have to write a default value everywhere
>
Most of programmers would like to do it. Explicitly defined logic is easier to read than some magic `-1` value that comes from C++ function to Lua.
Of course after XX years of `-1` as default value it looks obvious for OTS programmers, but for anyone outside OTS community it's unreadable code: why do they compare to -1 everywhere?

// offtop
When I've worked on Kasteria.pl, which was based on ~2010 'otserv' engine, it had 'getPlayerStorageValue' that returned 2 values from C++: storage value [or -1], is it defined ex.:
```lua
local value, defined = getPlayerStorageValue(123)
```
I think that version with default value as second parameter to `getStorage` is better.